pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,345 @@
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			uids := []types.UID{}
+			for _, schemaName := range export.Spec.LatestResourceSchemas {
+				schema, err := apiResourceSchemaInformer.Lister().Get(client.ToClusterAwareKey(exportCluster, schemaName))
+
+				if errors.IsNotFound(err) {
+					schema, err = temporaryRemoteShardAPIResourceSchemaInformer.Lister().Get(client.ToClusterAwareKey(exportCluster, schemaName))
+				}
+
+				if err != nil {
+					return nil, err
+				}
+
+				uids = append(uids, schema.UID)
+			}
+			return uids, nil
+		},
+		deleteCRD: func(ctx context.Context, name string) error {
+			return crdClusterClient.ApiextensionsV1().CustomResourceDefinitions().Cluster(apibinding.ShadowWorkspaceName).Delete(ctx, name, metav1.DeleteOptions{})
+		},
+	}
+
+	indexers.AddIfNotPresentOrDie(
+		apiBindingInformer.Informer().GetIndexer(),
+		cache.Indexers{
+			indexers.APIBindingByBoundResourceUID: indexers.IndexAPIBindingByBoundResourceUID,
+		},
+	)
+
+	crdInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+
+			if !ok {
+				return false
+			}
+
+			return logicalcluster.From(crd) == apibinding.ShadowWorkspaceName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueCRD(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+		},
+	})
+
+	apiBindingInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.enqueueFromAPIBinding(oldObj, newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueFromAPIBinding(obj, obj)
+			},
+		},
+	)
+
+	return c, nil
+}
+
+// controller deletes bound CRDs when they are no longer in use by any APIBindings
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	getCRD                           func(clusterName logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error)
+	getAPIBindingsByBoundResourceUID func(name string) ([]interface{}, error)
+	listAPIBindings                  func() ([]*apisv1alpha1.APIBinding, error)
+	getResourceUIDs                  func(binding *apisv1alpha1.APIBinding) ([]types.UID, error)
+	deleteCRD                        func(ctx context.Context, name string) error
+}
+
+// enqueueCRD enqueues a CRD.
+func (c *controller) enqueueCRD(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing CRD")
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueFromAPIBinding(oldObj, newObj interface{}) {
+	oldBinding, ok := oldObj.(*apisv1alpha1.APIBinding)
```

Comment:
this looks wrong. On add this cast will fail.

Reply:
The behavior is without consequence, but must be very explicit here with a nil check to be able to reason about the code. Now it looks like an conincidence.

Reply:
`oldObj` can be `nil` now, and use of it in this function is wrapped in a nil check.